### PR TITLE
Added support for delayed violation and provisioning enforcement

### PR DESCRIPTION
**Pull request by jrouzierinverse**

### CODE REVIEW

File: db/upgrade-4.4.0-4.5.0.sql [hunk 1]
```diff
@@ -0,0 +1,8 @@
+--
+-- Alter for dynamic controller
```
Comment (cgx) [inline db/upgrade-4.4.0-4.5.0.sql]: Wrong comment.

File: html/pfappserver/lib/pfappserver/Form/Violation.pm [hunk 2]
```diff
@@ -148,6 +148,13 @@ has_field 'window' =>
    tags => { after_element => \&help,
              help => 'Amount of time before a violation will be closed automatically. Instead of allowing people to reactivate the network, you may want to open a violation for a defined amount of time instead.' },
   );
+has_field 'delay_by' =>
+  (
+   type => 'Duration',
+   label => 'Delay By',
+   tags => { after_element => \&help,
+             help => "Amount of time a violation's will be triggered" },
```
Comment (cgx) [inline html/pfappserver/lib/pfappserver/Form/Violation.pm]: I would say: _Delay before triggering the violation._

File: html/pfappserver/lib/pfappserver/Controller/User.pm [hunk 3]
```diff
@@ -107,10 +107,6 @@ sub object :Chained('/') :PathPart('user') :CaptureArgs(1) {
     if (is_success($status)) {
         $c->stash->{user} = pop @{$result};
         # Fetch associated nodes
-        ($status, $result) = $self->getModel($c)->nodes($pid);
```
Comment (cgx) [inline html/pfappserver/lib/pfappserver/Controller/User.pm:110]: Good move :+1:

File: html/pfappserver/lib/pfappserver/I18N/en.po [hunk 4]
```diff
@@ -4382,6 +4382,10 @@ msgstr ""
 msgid "log_action"
 msgstr "Log message"
 
+# pf::action (VIOLATION_ACTIONS)
+msgid "enforce_provisioning_action"
+msgstr "Enforce Provisioning"
```
Comment (cgx) [inline html/pfappserver/lib/pfappserver/I18N/en.po:4387]: We'll need to make sure this string is extracted by `addons/extract_i18n_strings.pl`.

File: html/pfappserver/lib/pfappserver/Model/Node.pm [hunk 5]
```diff
@@ -505,6 +505,18 @@ sub closeViolation {
...
 }
 
+=head2 runViolation
+
+=cut
+
+sub runViolation {
+    my ($self, $id) = @_;
+    if(violation_run_delayed($id)) {
+        return ($STATUS::OK, 'The violation was successfully ran');
```
Comment (cgx) [inline html/pfappserver/lib/pfappserver/Model/Node.pm:515]: Do we run a violation or trigger a violation?

File: html/pfappserver/root/node/violations.tt [hunk 6]
```diff
@@ -11,10 +12,20 @@
     <tbody>
     [%- FOR item IN items %]
     <tr[% ' class="muted"' IF item.status != 'open' %]>
```
Comment (cgx) [inline html/pfappserver/root/node/violations.tt]: I would show _delayed_ violation as **not** muted. The test should be: `IF item.status == 'closed'`

File: html/pfappserver/root/node/violations.tt [hunk 7]
```diff
@@ -11,10 +12,20 @@
...
+      <td>[% item.status | html %]</td>
```
Comment (cgx) [inline html/pfappserver/root/node/violations.tt]: How about simply putting a [clock icon](http://fortawesome.github.io/Font-Awesome/icon/clock-o/) on the left of the description?

File: html/pfappserver/root/node/violations.tt [hunk 8]
```diff
@@ -11,10 +12,20 @@
...
       <td>[% item.description | html %]</td>
       <td>[% item.start_date | html %]</td>
       <td>[% item.release_date | html %]</td>
-      <td>[% IF item.status == 'open' && can_access("NODES_UPDATE") %]<a href="[% c.uri_for(c.controller('Node').action_for('closeViolation'), [item.id]) %]" class="btn btn-mini">[% l('Release') %]</a>[% END %]</td>
+      <td>
+        [% IF can_access("NODES_UPDATE") %]
+            [% IF item.status == 'open' %]
+                <a href="[% c.uri_for(c.controller('Node').action_for('closeViolation'), [item.id]) %]" class="btn btn-mini">[% l('Release') %]</a>
+            [% ELSIF item.status == 'delayed'%]
+                <a href="[% c.uri_for(c.controller('Node').action_for('runViolation'), [item.id]) %]" class="btn btn-mini">[% l('Run') %]</a>
```
Comment (cgx) [inline html/pfappserver/root/node/violations.tt]: I suggest to name this button **Trigger Now**, to make it clear that this is a delayed violation.

File: lib/pf/api.pm [hunk 9]
```diff
@@ -278,6 +278,21 @@ sub _reassignSNMPConnections {
     $logger->info( "[$mac] Flipping admin status on switch (".$switch->{'_id'}.") ifIndex $ifIndex. " );
     $switch->bouncePort($ifIndex);
 }
+
+
+=head2 violation_delayed_run
+
+TODO: documention
```
Comment (cgx) [inline lib/pf/api.pm]: Do it now!

File: lib/pf/client.pm [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+use Module::Pluggable search_path => 'pf::api', sub_name => 'modules', require => 1;
+
+my @MODULES = __PACKAGE__->modules;
+
+our $DEFAULT_CLIENT = "pf::api::jsonrpcclient";
+our $CURRENT_CLIENT = $DEFAULT_CLIENT;
+
+=head2 setClient
+
+TODO: documention
```
Comment (cgx) [inline lib/pf/client.pm]: Please ...

File: lib/pf/client.pm [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+sub setClient {
+    my ($clienttype) = @_;
+    return unless any { $_ eq $clienttype } @MODULES; 
+    $CURRENT_CLIENT= $clienttype;
+}
+
+=head2 getClient
+
+TODO: documention
```
Comment (cgx) [inline lib/pf/client.pm]: :facepunch:

File: lib/pf/violation_config.pm [hunk 12]
```diff
@@ -57,12 +57,13 @@ sub loadViolationsIntoDb {
...
 
         # parse grace, try to understand trailing signs, and convert back to seconds
-        if ( defined $data->{'grace'} ) {
-            $data->{'grace'} = normalize_time($data->{'grace'});
-        }
-
-        if ( defined $data->{'window'} && $data->{'window'} ne "dynamic" ) {
-            $data->{'window'} = normalize_time($data->{'window'});
+        my @time_values = (qw(grace delay_by));
+        push @time_values,'window' if defined $data->{'window'} && $data->{'window'} ne "dynamic";
```
Comment (cgx) [inline lib/pf/violation_config.pm]: Please put some parentheses, it will make the code more readable.

File: sbin/pfmon [hunk 13]
```diff
@@ -142,7 +142,7 @@ sub registertasks  {
     );
     register_task(
         'checking violations for expiration','violation_maintenance_interval',sub {
-            violation_maintenance();
+            violation_maintenance(1000,10);
```
Comment (cgx) [inline sbin/pfmon]: Magical numbers? Document and define constants please!

File: html/pfappserver/lib/pfappserver/Model/Node.pm (hunk 5 above)
Comment (rouzier) [inline html/pfappserver/lib/pfappserver/Model/Node.pm:515]: We run it

File: lib/pf/violation.pm [hunk 14]
```diff
@@ -761,24 +767,75 @@ Check if we should close violations based on release_date
...
+                # If close is a success, reevaluate the Access for the node
+                if ($result) {
+                    pf::enforcement::reevaluate_access( $mac, "manage_vclose" );
+                }
+            }
+        }
+        $query->finish;
+        $endtime = time;
+        $logger->trace("starttime: $start_time, endtime: $endtime");
+    } while( (($endtime - $start_time) < $timelimit) );
```
Comment (rouzier) [inline lib/pf/violation.pm:805]: It automatically is converted since it is marked as a time type in documentation.conf.

**Pull request merged**